### DISCUSSION TITLE
How can we use mock outputs when plans use a data resource that hasn't been applied yet?

### DISCUSSION THREAD

**Discussion by calebtote** [gruntwork-io/knowledge-base#589] [Help]
**Question:** If we are running a `plan` for a module that does a data lookup, and that resource doesn't yet exist, what options do we have for the first time we deploy?

We are referencing: https://github.com/gruntwork-io/terraform-aws-ecs/blob/08ec667c8dfbc84ca21471e8bfcbb1218408ade6/modules/ecs-service

Our terragrunt.hcl has:

```
dependency iam_role_ecs_task {
  config_path = "${path_relative_from_include()}/live/global/iam/ecs/${local.app}-role"

  mock_outputs = {
    role_name = "MOCK.iam.role"
  }
  mock_outputs_allowed_terraform_commands = ["validate", "plan"]
}

[...]

inputs = {
  ...
  existing_ecs_task_role_name           = dependency.iam_role_ecs_task.outputs.role_name
  ...
}
```

Because of [the data resource](https://github.com/gruntwork-io/terraform-aws-ecs/blob/08ec667c8dfbc84ca21471e8bfcbb1218408ade6/modules/ecs-service/task_definition.tf#L194) in the module we're referencing, our `terragrunt plan` jobs fail with:

```
│ Error: error reading IAM Role (MOCK.iam.role.name): NoSuchEntity: The role with name MOCK.iam.role.name cannot be found.
│       status code: 404, request id: 21aa89c0-1134-4048-a487-5cc1ffbf9c79
```

In `dev`, our solution for fresh deploys is to just apply the iam-role dependency first and then everything works as expected. That's not a great solution for `stable` or `PRODUCTION` though. What's the correct way to have our plans succeed here, prior to running apply (without a double apply)?

---

<ins datetime="2022-11-09T17:29:40Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109578">Tracked in ticket #109578</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
since reading of outputs from plan is not supported it will fail with errors during `terragrunt plan`

https://terragrunt.gruntwork.io/docs/features/execute-terraform-commands-on-multiple-modules-at-once/#unapplied-dependency-and-mock-outputs

**Reply by calebtote** [reply]
Right, I think we understand that much so far - I'm trying to determine if there's a way to get around this without a double-apply where we have to first apply the dependency and then apply everything else. For context, in this scenario we're running a `run-all plan` followed by a `run-all apply` -- I want to avoid needing to `plan` and `apply` specific resources for our `run-all` to succeed the first time.
